File: Exercise2/threads.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>


int i = 0;
pthread_mutex_t lock;

void* thread_1Function() {

	int x = 0;
	while(x < 1000000){
		pthread_mutex_lock(&lock);
		i++;
		x++;
		pthread_mutex_unlock(&lock);
	}

	return NULL;
}

void* thread_2Function() {

	int x = 0;
	while(x < 1000000){
		pthread_mutex_lock(&lock);
		i--;
		x++;
		pthread_mutex_unlock(&lock);
	}

	return NULL;
}




int main() {
	pthread_t thread_1;
	pthread_create(&thread_1, NULL, thread_1Function, NULL);

	pthread_t thread_2;
	pthread_create(&thread_2, NULL, thread_2Function, NULL);

	pthread_join(thread_1, NULL);
	pthread_join(thread_2, NULL);

	printf("%d\n", i);

	pthread_mutex_destroy(&lock);
	return 0;
}

File: Exercise2/threads.py
from threading import Thread
import threading

i = 0
lock = threading.Lock()
	


def thread_1_Function():
	global i
	for x in range(0,1000001):
		lock.acquire()
		i+=1
		lock.release()
	

def thread_2_Function():
	global i	
	for x in range(0,1000000):
		lock.acquire()
		i-=1
		lock.release()


def main():
	thread_1 = Thread(target = thread_1_Function, args = (),)
	thread_2 = Thread(target = thread_2_Function, args = (),)
	
	thread_1.start()
	thread_2.start()

	thread_1.join()
	thread_2.join()
	
	print(i)

main()
